### PR TITLE
Move linting and unit tests to Github Actions

### PR DESCRIPTION
**Pull request by jayofdoom**
This PR migrates our unit tests and linting to github actions. It uses some of the prior art from PR #675 as a guide, but was created greenfield otherwise.

The testing matrix is equivalent to our matrix in CircleCI -- we're testing both against go 1.16 and go 1.18.

### CODE REVIEW

**Comment by jayofdoom** [comment]
Successful test passes can be seen in my fork:
https://github.com/jayofdoom/armada/actions/runs/2423754952
https://github.com/jayofdoom/armada/actions/runs/2423754957

File: .github/workflows/go.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+on: [push]
```
Comment (kannon92) [inline .github/workflows/go.yml:1]: You can add more options to this.  I think push is good.  

I've seen people have pull_request as one option also.  

I'm not sure if you need to change it though.  I've seen people use pretty elaborate rules but we aren't there yet.

**Review by kannon92 — APPROVED** [review]

File: .github/workflows/go.yml (hunk 1 above)
Comment (jayofdoom) [reply .github/workflows/go.yml:1]: I have intentionally not pursued things like this yet, including e.g. only running the TS linter if TS files are changed... right now I'm trying to stick to replicating functionality provided by CircleCI.

**Comment by jayofdoom** [comment]
Whoever merges this will have to mash buttons to change the required jobs for merge to remove the circleci check_and_unit_test_* required ones, and instead require the workflow to pass.

File: .github/workflows/go.yml (hunk 1 above)
Comment (kannon92) [reply .github/workflows/go.yml:1]: What I was wondering is for github actions, will they run when a PR is merged into main?  

It says push but is a merge considered a push or a pull_request closed?

I'm not familar with circle CI so I don't know if circle CI has sensible defaults.

Comment (jayofdoom) [reply .github/workflows/go.yml:1]: That's a solid question, I'll check that behavior.

Comment (jayofdoom) [reply .github/workflows/go.yml:1]: https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#push

A merged pull request will trigger the push webhook. So it'll run on push.

(Just saying this out loud) -- We'll need to be more delicate about what we run jobs on once we add releasing to the github actions.

**Pull request merged**